glean-core/ffi/src/timing_distribution.rs

```diff
@@ -0,0 +1,126 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::convert::TryFrom;
+use std::os::raw::c_char;
+
+use ffi_support::{define_handle_map_deleter, ConcurrentHandleMap, FfiStr};
+use lazy_static::lazy_static;
+
+use glean_core::metrics::{MetricType, TimeUnit, TimingDistributionMetric};
+use glean_core::{CommonMetricData, Lifetime};
+
+use crate::handlemap_ext::HandleMapExtension;
+use crate::{from_raw_string_array, RawStringArray, GLEAN};
+
+lazy_static! {
+    pub static ref TIMING_DISTRIBUTION_METRICS: ConcurrentHandleMap<TimingDistributionMetric> =
+        ConcurrentHandleMap::new();
+}
+define_handle_map_deleter!(
+    TIMING_DISTRIBUTION_METRICS,
+    glean_destroy_timing_distribution_metric
+);
+
+#[no_mangle]
+pub extern "C" fn glean_new_timing_distribution_metric(
+    category: FfiStr,
+    name: FfiStr,
+    send_in_pings: RawStringArray,
+    send_in_pings_len: i32,
+    lifetime: i32,
+    disabled: u8,
+    time_unit: i32,
+) -> u64 {
+    TIMING_DISTRIBUTION_METRICS.insert_with_log(|| {
+        let send_in_pings = unsafe { from_raw_string_array(send_in_pings, send_in_pings_len) };
+        let lifetime = Lifetime::try_from(lifetime)?;
+        let tu = TimeUnit::try_from(time_unit)?;
+
+        Ok(TimingDistributionMetric::new(
+            CommonMetricData {
+                name: name.into_string(),
+                category: category.into_string(),
+                send_in_pings,
+                lifetime,
+                disabled: disabled != 0,
+            },
+            tu,
+        ))
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn glean_timing_distribution_should_record(glean_handle: u64, metric_id: u64) -> u8 {
+    GLEAN.call_infallible(glean_handle, |glean| {
+        TIMING_DISTRIBUTION_METRICS
+            .call_infallible(metric_id, |metric| metric.should_record(&glean))
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn glean_timing_distribution_set_start(
```

Comment:
I was a bit disappointed that we have to provide the API to set the start timestamp as an integer.
I do get now why we need that instead of using `Instant::now` internally.

The reason why I went with the `set_*` prefix for start and stop is that I _wanted to keep just `start` and `stop` open for the proper rust api that _does_ use `Instant`.
(We have the same for the Timespan)

Reply:
Ah, I see, thanks for explaining :) Makes sense!